Framework/Beamline/inc/MantidBeamline/ComponentInfo.h

```diff
@@ -39,19 +41,51 @@ class MANTID_BEAMLINE_DLL ComponentInfo {
 
 private:
   boost::shared_ptr<const std::vector<size_t>> m_assemblySortedDetectorIndices;
-  boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>> m_ranges;
-  size_t m_size = 0;
+  boost::shared_ptr<const std::vector<size_t>> m_assemblySortedComponentIndices;
+  /// Ranges of component ids that are contiguous blocks of detectors.
+  boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>>
+      m_detectorRanges;
+  /// Ranges of component ids that are contiguous blocks of components.
+  boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>>
+      m_componentRanges;
+  Mantid::Kernel::cow_ptr<std::vector<Eigen::Vector3d>> m_positions;
+  Mantid::Kernel::cow_ptr<std::vector<Eigen::Quaterniond>> m_rotations;
+  const size_t m_size = 0;
+  boost::shared_ptr<DetectorInfo> m_detectorInfo;
 
 public:
   ComponentInfo();
-  ComponentInfo(
-      boost::shared_ptr<const std::vector<size_t>>
-          assemblySortedDetectorIndices,
-      boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>> ranges);
+  ComponentInfo(boost::shared_ptr<const std::vector<size_t>>
+                    assemblySortedDetectorIndices,
+                boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>>
+                    detectorRanges,
+                boost::shared_ptr<const std::vector<size_t>>
+                    assemblySortedComponentIndices,
+                boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>>
+                    componentRanges,
+                boost::shared_ptr<std::vector<Eigen::Vector3d>> positions,
+                boost::shared_ptr<std::vector<Eigen::Quaterniond>> rotations,
+                boost::shared_ptr<DetectorInfo> detectorInfo);
+
   std::vector<size_t> detectorIndices(const size_t componentIndex) const;
+  std::vector<size_t> componentIndices(const size_t componentIndex) const;
```

Comment:
I think I would expect having the original index included -- if it was not included the result would be a bush, not a tree.

For consistency, should then also `detectorIndices` be renamed?

```cpp
detectorsInSubtree();
componentsInSubtree();
```

I am not 100% happy with this name either, but I feel it is slightly better. Maybe the main decision is whether we want to keep referring to "index" and "indices" all the time? Basically, is it obvious what code like this does:

```cpp
for (const auto &det : componentInfo.detectorsInSubtree(comp)) {
  detectorInfo.setPosition(det, Vector3d(0,0,0));
}
```

Woult it be less or more readable like this:

```cpp
for (const auto &detIndex : componentInfo.detectorsIndices(compIndex)) {
  detectorInfo.setPosition(detIndex, Vector3d(0,0,0));
}
```

To keep this consistent, we might then also consider renaming the `indexOf` methods?